feat: display and sign order with partner fee

shoom3301 opened the pull request:
# Summary

1. Fixed price details accordion displaying (display it only we have a valid trade)
2. Added a total fee amount on the collapsed accordion state (fee + partnerFee)
3. Made the accordion open by default on the confirm screen
4. Changed "Receive (inc. fee)" displayed value and added partner fee into tooltip (sell/buy orders)
5. Removed price info duplication from the accordion
6. Add partner fee into "Expected to receive" and "Minimum receive" values on the confirm screen
7. Added partner fee into appData (it actually was already there, I just fixed the mocking data)

<img width="486" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/944337d5-350c-48b3-9fde-65816822e883">

<img width="474" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/e706bdf3-62e9-4298-a747-8632587603a1">

<img width="420" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/dd80660d-cc2c-4afe-a2c4-baa558d61cb2">

<img width="423" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/ecba47d8-02ae-4939-a376-31bd22290b2e">

<img width="511" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/680ace71-a008-4eb8-97f9-d3e1d75b1543">

  # To Test

To avoid confusion, please test through https://github.com/cowprotocol/cowswap/pull/3942
Because there I've deleted non-zero-fee mode, so now we always use fee=zero for swaps.

>In this PR partner-fee is hardcoded as 25%

>https://github.com/cowprotocol/cowswap/pull/3943 - here you can test the changes without mocking partner fee

1. Sell/buy orders
- [ ] "Receive (inc. fee)" should include fee + partnerFee values
- [ ] "From (inc. fee)" should include fee + partnerFee values
2. Confirmation screen
- [ ] "Expected to receive" should include fee + partnerFee values
- [ ] "Minimum receive" should include fee + partnerFee values
- [ ] Sining data should match "Minimum receive"
3. PartnerFee mush be included into appData

shoom3301 commented on apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.ts:
```diff
@@ -24,14 +24,14 @@ function inputAmountForSignature(params: AmountForSignatureParams): CurrencyAmou
+    return trade.inputAmountWithPartnerFee
   } else {
-    return trade.inputAmountWithoutFee.multiply(slippageCoeff) // add slippage
+    return trade.inputAmountWithPartnerFee.multiply(slippageCoeff) // add slippage
```
I'm not sure here

shoom3301 commented on apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:
```diff
@@ -86,8 +106,14 @@ describe('getAmountsForSignature()', () => {
-      // inputAmount = (3012) + 5% = 3,162.6
-      expect(result.inputAmount.toFixed()).toEqual('3162.600000')
+      if (partnerFeeBps === 0) {
+        // inputAmount = (3012 + 8) + 5% = 3171.000000
```
I think we should not add fee here

shoom3301 commented on apps/cowswap-frontend/src/modules/injectedWidget/hooks/useInjectedWidgetParams.ts:
```diff
@@ -11,7 +11,7 @@ export function useInjectedWidgetParams(): Partial<CowSwapWidgetParams> {
     ...value,
     // TODO: Remove the default partner fee after testing
```
TODO: remove the mock

shoom3301 replied:
https://github.com/cowprotocol/cowswap/pull/3943

alfetopito commented on apps/cowswap-frontend/src/legacy/state/swap/extension.ts:
```diff
@@ -55,6 +54,8 @@ export function useTradeExactInWithFee({
 
   // external price output as Currency
   const outputAmount = stringToCurrency(quote.price.amount, outputCurrency)
+  const partnerFeeAmount = partnerFee ? outputAmount.multiply(bpsToPercent(partnerFee.bps)) : undefined
```
Can we use `0` for `partnerFeeAmount`, so there's no need for conditional and it's safe to always subtract it from `outputAmount`?

shoom3301 replied:
I would prefer to keep it more explicit.
If we set 0 here, then it will be difficult to distinguish if the partner fee is not set or set to 0

anxolin commented on apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:
```diff
@@ -67,17 +81,23 @@ describe('getAmountsForSignature()', () => {
+        // outputAmount = 2 - 5% = 1.900000000000000000
+        expect(result.outputAmount.toFixed()).toEqual('1.900000000000000000')
+      } else {
+        // outputAmount = 2 - 5% - 20% = 1.520000000000000000
```
this is a bit hard to follow, as you are assuming you get a 20% because there's only two tests and one is zero, and the other 20%.

One other way, could be to add this in the `describe.each`

what about something like:

```js
describe.each([
   {partnerFeeBps: 0, expectedOutputAmount: '1.900000000000000000' }, //  explainer here...
   {partnerFeeBps: 2000, expectedOutputAmount: '1.520000000000000000' }, //  explainer here...
])
```

anxolin replied:
In the above example, fee is constant. I think is good, as these test would focus on testing variations on the partnerFee for sell orders, but we will need some that focus on variations on the fee too.

Alternativelly, you might want to do a mega-test with all your cases:

describe.each([
   {partnerFeeBps: 0,  fee: 0, slippage: 0, expectedOutputAmount: '1.900000000000000000' }, //  explainer here...
   ... kee some params constants and change the others, to assert the result given some combinations
])

shoom3301 replied:
Refactored the tests, split them into separate cases in WHEN/THEN style

anxolin commented on apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:
```diff
@@ -114,16 +142,23 @@ describe('getAmountsForSignature()', () => {
+      } else {
+        // outputAmount = 2 - 5% - 20% = 1.520000000000000000
+        expect(result.outputAmount.toFixed()).toEqual('1.520000000000000000')
+      }
```
same

shoom3301 replied:
Also refaactored

anxolin reviewed (commented):
This is a great PR!!!!

The labels will change a lot, so feel free to ignore this so we tackle all things together, but I believe we will need to signal this number is an estimation only:

<img width="659" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/a7f441f5-f7d6-464c-9518-87a72bc5508c">

alfetopito commented on apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:
```diff
@@ -55,7 +67,9 @@ describe('getAmountsForSignature()', () => {
-      const trade = getTrade({ tradeType: TradeType.EXACT_INPUT, inputAmount, outputAmount, feeAmount })
+      const trade = getTrade({ tradeType: TradeType.EXACT_INPUT, inputAmount, outputAmount, feeAmount, partnerFeeBps })
+
+      if (!trade) throw new Error('No trade')
```
This test with several conditionals don't look good to me.
Shouldn't we asset the trade exists and fail if it doesn't?

shoom3301 replied:
This is a part of legacy comes from 'buildTradeExactInWithFee'.
I would make it returning always a trade with no falsy values, but it will require significant refactoring now.
Don't you mind to keep it for the future?

alfetopito commented on apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:
```diff
@@ -67,17 +81,23 @@ describe('getAmountsForSignature()', () => {
-      // Fee in WETH = 0.005312084993359893
-      // outputAmount = (2 - 0.005312) - 5% = 1.89495
-      expect(result.outputAmount.toFixed()).toEqual('1.894953519256308100')
+      if (partnerFeeBps === 0) {
+        // outputAmount = 2 - 5% = 1.900000000000000000
+        expect(result.outputAmount.toFixed()).toEqual('1.900000000000000000')
+      } else {
+        // outputAmount = 2 - 5% - 20% = 1.520000000000000000
+        expect(result.outputAmount.toFixed()).toEqual('1.520000000000000000')
+      }
```
Here too. These are 2 different tests to me.

shoom3301 replied:
Fixed

alfetopito commented on apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts:
```diff
@@ -28,6 +28,8 @@ export const defaultCurrencyInputPanelProps: CurrencyInputPanelProps & { priceIm
       amountAfterFeesRaw: CurrencyAmount.fromRawAmount(currency, 20 * 10 ** 18),
       feeAmount: '10',
       feeAmountRaw: CurrencyAmount.fromRawAmount(currency, 10 * 10 ** 18),
+      partnerFeeAmount: '25',
+      partnerFeeAmountRaw: CurrencyAmount.fromRawAmount(currency, 25 * 10 ** 18),
```
Why does this default to non-zero amounts?

shoom3301 replied:
Changed to 0

elena-zh commented:
Hey @shoom3301 , this is a great PR!!
(I tested it in #3942) 
However, there are some issues I faced: 
1. There is no fee amount (min) when fees exceed from amount error is displayed (see the comparison with prod)
![fee amount](https://github.com/cowprotocol/cowswap/assets/70885163/b4b2745e-af50-4d86-965c-a5be95a5a7fb)
![fees example prod](https://github.com/cowprotocol/cowswap/assets/70885163/3012b2a5-3e6b-4fc0-90a7-597d649500fd)
2. There is significant difference in a partner fee calculation for a sell and a buy orders. See details in https://cowservises.slack.com/archives/C0361CDG8GP/p1709127531498219
3. Fee threshold banners should take into account a partner fee (see in https://cowservises.slack.com/archives/C0361CDG8GP/p1709124571171179)
4. I'd add '+gas' into the fee field when an order is onchain (SC wallets, ETH orders) (see https://cowservises.slack.com/archives/C0361CDG8GP/p1709126523835639)
5. when no USD estimation, we decided to have fees in a token value instead of USD estimation 
![value in tokens](https://github.com/cowprotocol/cowswap/assets/70885163/9af932f4-c93f-48d5-ac84-95596e5919a8)
6. It would be great to enhance values displaying in a mobile view 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/7902e246-e1a4-4093-a9c3-d648549c4c6d)


Thank you!

shoom3301 commented:
@elena-zh thank you!
1. Fixed
2. Fixed
3. Fixed
4. Sorry, I'm not sure how it should looks like. I would keep it for the future, we need Michel here
5. Fixed
6. Fixed only arrow color, I'm not sure about the reset, we also need Michel here

elena-zh commented:
Hey @shoom3301 , great!

For the 6th case, I have opened #3953 issue for @fairlighteth to take a look at. 
For the case 4, I found 2 already opened tasks on the board (#1760, #2033)

As for the fixes: 
1. Case 1: could you please add a label to the collapsed dropdown 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/0e4954a7-4d51-46a8-892f-f4348c84d95d)
2. And I can see a negative amount to receive for some trades :( 
![receive minus](https://github.com/cowprotocol/cowswap/assets/70885163/f92a320f-57c6-4f6f-9b62-e832c4ed8e13)

Could you please fix this?

shoom3301 commented on apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts:
```diff
@@ -28,6 +28,8 @@ export const defaultCurrencyInputPanelProps: CurrencyInputPanelProps & { priceIm
       feeAmount: '10',
       feeAmountRaw: CurrencyAmount.fromRawAmount(currency, 10 * 10 ** 18),
+      partnerFeeAmount: '25',
+      partnerFeeAmountRaw: CurrencyAmount.fromRawAmount(currency, 0),
```
This object is a mock for Cosmos.
I've specified values because they are required by the component.

p.s. changed value of partnerFeeAmount to 0

alfetopito replied:
Ah ok, if this is for cosmos it makes sense.
During the review and based on the file name it's not clear this is mock data.

shoom3301 commented on apps/cowswap-frontend/src/mocks/tradeStateMock.ts:
```diff
@@ -39,6 +41,8 @@ export const outputCurrencyInfoMock: CurrencyInfo = {
     feeAmount: '10',
     feeAmountRaw: CurrencyAmount.fromRawAmount(outputCurrency, 10 * 10 ** 18),
+    partnerFeeAmount: '25',
+    partnerFeeAmountRaw: CurrencyAmount.fromRawAmount(outputCurrency, 0),
```
https://github.com/cowprotocol/cowswap/pull/3941#discussion_r1508528292

shoom3301 commented:
@elena-zh thank you! Fixed both

elena-zh commented:
Hey @shoom3301 , case 1 works great!
As for the case 2, I see calculation mismatch in the tooltip and in the fees dropdown :(
![image](https://github.com/cowprotocol/cowswap/assets/70885163/0718bcbe-bf49-49c8-9b6f-475fed697b80)

shoom3301 commented:
@elena-zh fixed

anxolin commented on apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts:
```diff
@@ -30,16 +30,16 @@ function useAgnosticPermitDataIfUserHasNoAllowance(): PermitHookData | undefined
 }
 
 export function AppDataHooksUpdater(): null {
-  const { v2Trade } = useDerivedSwapInfo()
```
great!, i hate to see GP trades and gpv2 stuff

elena-zh commented:
@shoom3301 , still I can face some issues: 
1. Sometimes I can't see estimated fee amount when sell amount is small (see 2 screenshots)
![image](https://github.com/cowprotocol/cowswap/assets/70885163/c5d90e8c-4c2c-4a44-92b2-5f1d6c1f89ff)
![image](https://github.com/cowprotocol/cowswap/assets/70885163/208d4f5e-fb06-4f83-b447-0ae696046082)
2. There is an error in percentage calculation for buy orders in the 'Fees exceed' banner
![image](https://github.com/cowprotocol/cowswap/assets/70885163/2e21e367-5e59-484b-8a09-c8b824c27dc9)
According to the formula on Prod, it is calculated based on the From (incl/fee) amount

Could you please fix this?

shoom3301 commented:
@elena-zh thank you!

1. This is not a new issue, it's actually this one: https://cowservices.slack.com/archives/C0361CDG8GP/p1709035605377909. I would like to fix it later, since it requires more analysis and work
2. Fixed

elena-zh commented:
Hey @shoom3301 , thank you!
However, 
1. The issue here is that est fees field is no longer displayed in some cases, however, it should. And it does not appear when I wait for a while
![est fees field](https://github.com/cowprotocol/cowswap/assets/70885163/3821f389-9b0f-4543-bb27-d2b7950be943)
2. I think that still there is something wrong with the percentage calculation. See:
![percentage](https://github.com/cowprotocol/cowswap/assets/70885163/06973bad-ac9a-4181-a3c6-27fdcd45d7ae)

shoom3301 commented:
@elena-zh thank you!
Fixed both

elena-zh reviewed (approved):
Approved, thank you!
